controllers/cluster_controller_phases.go

```diff
@@ -88,13 +88,14 @@ func (r *ClusterReconciler) reconcileExternal(ctx context.Context, cluster *clus
 		UID:        cluster.UID,
 	}
 
-	if !util.HasOwnerRef(obj.GetOwnerReferences(), ownerRef) {
-		obj.SetOwnerReferences(util.EnsureOwnerRef(obj.GetOwnerReferences(), ownerRef))
-		if err := r.Client.Patch(ctx, obj, objPatch); err != nil {
-			return nil, errors.Wrapf(err,
-				"failed to set OwnerReference on %v %q for Cluster %q in namespace %q",
-				obj.GroupVersionKind(), ref.Name, cluster.Name, cluster.Namespace)
-		}
+	// Set the owner reference. EnsureOwnerRef will add ownerRef to a slice of
```

Comment:
I'm not sure these comments are necessary now. I don't mind adding them but I believe this code does not need comments.

Do you have a preference here @vincepri?

Reply:
The comment can be removed, no strong preferences here.

That said, reading through the changes a little bit, it looks like we're now patching regardless if we need to add the OwnerReference or not, which I think it's a regression. IIRC we also shouldn't issue empty patches, which might lead us into errors. It might be worth to switch to @detiber's util/patch helper, which I think it actually checks if the patch has content before issuing it

Reply:
that makes sense. The `if !util.HasOwnerRef()` is also done inside of [`util.EnsureOwnerRef()`](https://github.com/kubernetes-sigs/cluster-api/blob/master/util/util.go#L281) so we would get rid of an iteration through all owner references. However we would be "recreating" all OwnerReferences each time this was called [k8s.io/apimachinery/pkg/api/meta/meta.go#L629](https://github.com/kubernetes/apimachinery/blob/master/pkg/api/meta/meta.go#L629).

So should I close this PR and make comment why on the issue @vincepri @chuckha ?

Reply:
Oh, I didn't realize empty patches lead to errors. Are you talking about connection errors or do empty patches somehow cause invalid data errors?

Reply:
Also, where can we find @detiber 's util/patch helpers for this?

Reply:
@alejandrox1 You can find the [patch helper here](https://github.com/kubernetes-sigs/cluster-api/blob/44491e485ce206f8e5203f858e9f8672b0763168/util/patch/patch.go) and you can see it being used https://github.com/kubernetes-sigs/cluster-api/blob/e214871fcc931e5f725bca62a62cb7209e669500/controllers/machine_controller.go#L87

Reply:
@vincepri my question seems to have gotten lost -- are you referring to http connection errors or is there a chance of some kind of data loss here or were you talking about other errors?

Reply:
There was an issue where an empty patch would trigger a reconciliation request regardless of changes, which we should avoid

Reply:
https://github.com/kubernetes-sigs/controller-runtime/issues/565

Reply:
Ok, that makes sense, thanks for the link.

@alejandrox1 The update to use the patch.NewHelper is the right approach here, whenever you add that feel free to ping me for another review!

Reply:
Thank you for all your comments!
I gave it another try, ptal
/cc @chuckha @vincepri @mcristina422